include/lsst/partition/ObjectIndex.h

```diff
@@ -0,0 +1,195 @@
+/*
+ * LSST Data Management System
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+/// \file
+/// \brief Operations with the "secondary" index
+
+#ifndef LSST_PARTITION_OBJECTINDEX_H
+#define LSST_PARTITION_OBJECTINDEX_H
+
+// System headers
+#include <fstream>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <mutex>
+#include <stdexcept>
+#include <string>
+
+// LSST headers
+#include "lsst/partition/Csv.h"
+
+// Forward declarations
+namespace lsst {
+namespace partition {
+    class ChunkLocation;
+}} // namespace lsst::partition
+
+// This header declarations
+namespace lsst {
+namespace partition {
+
+/**
+ * Class ObjectIndex is a singleton serving as a front-end to a specific implementation of
```

Comment:
A lot of things are bad for my health ... but they work, especially in a context of this command-line tool. I don't see how a global object would help to improve this particular application.

I have a strong desire to get rid of the overcomplicated "map-reduce" machinery and replace it with a much simple "parallelization logic. One of the benefits of that would be a possibility to directly pass a pointer of the "secondary index" interface to the partitioner instead of playing the nasty games (static, global) of the dependency injection. That  refactoring should be done in a separate ticket.

By the way, if you think more on whats wrong with the singleton (static or global) approach then you would probably realize that the singleton would prevent embedding the partitioner (as a library) into some "host" application. The "hoist" application may want to instantiate more than one partitioner. For instance, as I already mentioned it earlier I see a lot of value in having the partitioning done internally by the Ingest system, not by users. That would save us (as the data operations team) from having many issues with user-partitioned data products.

Reply:
The second item in my "wish list" for this application is to turn it into a library configured via `JSON`.